WordPress/gutenberg
# Fix author information leakage by author blocks for Custom Post Types without author support & display notice to user

fixes: #65816
## What?
This PR updates the behavior of author blocks to ensure they do not render when the post belongs to a Custom Post Type (CPT) that does not explicitly declare support for the author feature.

## Why?
Author blocks displaying on CPTs without declared author support is inconsistent with WordPress behavior.
This can lead to unintentional data exposure, as WordPress still saves author information to the database even when it is not intended to be visible.
Ensures better security, consistency, and adherence to user expectations for CPTs.

## How?

- Added a conditional check in the rendering logic of the author blocks to verify if the CPT supports the author feature.
- If the feature is unsupported, a message is displayed in the editor notifying users that the block will not render on the frontend.
- Updated the frontend output to prevent leaking author information when the CPT does not support authors.

## Testing Instructions
Reproduction:-
1. Create a Custom Post Type (CPT) without declaring support for the author feature:

```php
add_action( 'init', function() {
    register_post_type( 'custom-post', array(
        'public' => true,
        'supports' => array( 'title', 'editor', 'custom-fields' ),
    ) );
} );
```

2. Add a post to the newly created CPT. 
3. Add an author-related block (e.g., Post Author) in the block editor for that CPT post.
4. View the post on the frontend.

Expected Outcome:-

1. In the Editor:
    - A message is displayed informing users that the author block will not render because the CPT does not support the author feature.

2. On the Frontend:
    - The author block does not render, and no author information is displayed.

## Screenshots or screencast 

https://github.com/user-attachments/assets/0b8bde8b-b032-44bc-b3da-03b04f8cc441


The suggestion given by @dhruvang21 is also used in this PR from the [comment](https://github.com/WordPress/gutenberg/issues/65816#issuecomment-2396080596)
